### PR TITLE
Create SpendableOutputs events no matter the chain::Confirm order

### LINKED ISSUE #962: `ChannelMonitor::best_block_updated` before `ChannelMonitor::transactions_confirmed` may needlessly delay broadcasting by a block
We had a user who pointed out that we weren't broadcasting when we should have been after they called `ChannelMonitor::best_block_updated` with a block well after a CSV locktime and then called `ChannelMonitor::transactions_confirmed` with the transaction which we should have been spending (with a block height/hash a ways in the past). We would (I believe) have broadcasted the correct spending transaction after the next `best_block_updated` call, but we really shouldn't wait until then to broadcast if we get notified of an old transaction after its CSV is up.

### PR DESCRIPTION
**Pull request by TheBlueMatt**
We had a user who pointed out that we weren't creating
`SpendableOutputs` events when we should have been after they
called `ChannelMonitor::best_block_updated` with a block well
after a CSV locktime and then called
`ChannelMonitor::transactions_confirmed` with the transaction which
we should have been spending (with a block height/hash a ways in
the past).

This was due to `ChannelMonitor::transactions_confirmed` only
calling `ChannelMonitor::block_confirmed` with the height at which
the transactions were confirmed, resulting in all checks being done
against that, not the current height.

Further, in the same scenario, we also would not fail-back and HTLC
where the HTLC-Timeout transaction was confirmed more than
ANTI_REORG_DELAY blocks ago.

To address this, we use the best block height for confirmation
threshold checks in `ChannelMonitor::block_confirmed` and pass both
the confirmation and current heights through to
`OnchainTx::update_claims_view`, using each as appropriate.

Note to reviewers: please carefully check the `update_claims_view` part of the changeset here, there may very well be some unintended double-use of the height which wasn't properly separated (though I don't see any immediately).

Fixes #962.

### CODE REVIEW

File: lightning/src/ln/functional_tests.rs [hunk 1]
```diff
@@ -9262,3 +9261,85 @@ fn test_invalid_funding_tx() {
...
+	nodes[2].node.peer_disconnected(&nodes[1].node.get_our_node_id(), false);
+
+	nodes[1].node.force_close_channel(&channel_id).unwrap();
+	check_closed_broadcast!(nodes[1], true);
+	check_added_monitors!(nodes[1], 1);
+	let node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().split_off(0);
+	assert_eq!(node_txn.len(), 1);
+
+	let conf_height = nodes[1].best_block_info().1;
+	if !check_no_broadcast_too_early {
```
Comment (jkczyz) [inline lightning/src/ln/functional_tests.rs:9298]: The double negative is a little confusing. Maybe name the variable independent of what is being checked and instead for what action or scenario is being tested.

File: lightning/src/chain/channelmonitor.rs [hunk 2]
```diff
@@ -2008,33 +2008,37 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
...
+		fee_estimator: &F,
+		logger: &L,
 	) -> Vec<TransactionOutputs>
 	where
 		B::Target: BroadcasterInterface,
 		F::Target: FeeEstimator,
 		L::Target: Logger,
 	{
-		let should_broadcast = self.would_broadcast_at_height(height, &logger);
+		let should_broadcast = self.would_broadcast_at_height(self.best_block.height(), logger);
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:2033]: Hmm, suppose we could, I thought renaming height to `conf_height` was pretty clear (though I could spell it out). For `OnchainTx` I could tweak it if you think it makes more sense (given it doesn't have its own `best_block` field)?

[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/chain/channelmonitor.rs:2033]: Seems like there are a few scenarios:
- it doesn't need to be passed at all because the method already can access it through `self`, so why bother passing it? (e.g., remove `would_broadcast_at_height` can be named simply `should_broadcast` and doesn't need to take a height, nor does `get_broadcasted_holder_claims`
- only the current height is needed, so passing `&BestBlock` makes it explicit and prevents passing the wrong height (e.g., `has_reached_confirmation_threshold`)
- both heights are needed in `OnchainTx` (maybe leave as is since there is a calculation involved at the call site)

[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:2033]: I believe `get_broadcasted_holder_claims` needs the confirmation height, or at least appears to given it passes it to `PackageTx::build_package` as `height_original`, though its API seems like a good opportunity for cleanup. Did the rest.

[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/chain/channelmonitor.rs]: Hmmm... I see but you're calling `get_broadcasted_holder_claims` with the best block height in `block_confirmed`, right?

File: lightning/src/chain/channelmonitor.rs [hunk 3]
```diff
@@ -2008,33 +2008,37 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			self.is_paying_spendable_output(&tx, height, &logger);
 		}
 
-		self.block_confirmed(height, txn_matched, watch_outputs, claimable_outpoints, broadcaster, fee_estimator, logger)
+		self.block_confirmed(height, txn_matched, watch_outputs, claimable_outpoints, &broadcaster, &fee_estimator, &logger)
 	}
 
+	/// Update state for new block(s)/transaction(s) confirmed. Note that the caller *must* update
+	/// `self.best_block` before calling, and set `conf_height` to the height at which any new
```
Comment (jkczyz) [inline lightning/src/chain/channelmonitor.rs:2015]: Is this right? It isn't the case when called by `transactions_confirmed` as of now.

Comment (jkczyz) [inline lightning/src/chain/channelmonitor.rs:2015]: Remove the comma as it indicates a compound sentence but this is a compound predicate.

File: lightning/src/chain/channelmonitor.rs [hunk 4]
```diff
@@ -2008,33 +2008,37 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
...
+	/// transaction(s)/block(s) were confirmed at, even if it is not the current best height
+	/// (assuming we are being called with new transaction(s) confirmed).
```
Comment (jkczyz) [inline lightning/src/chain/channelmonitor.rs:2017]: The parenthetical seems unnecessary and a little hard to follow. I'd suggest dropping it.

File: lightning/src/chain/onchaintx.rs [hunk 5]
```diff
@@ -377,12 +377,15 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
 	/// for this channel, provide new relevant on-chain transactions and/or new claim requests.
 	/// Formerly this was named `block_connected`, but it is now also used for claiming an HTLC output
 	/// if we receive a preimage after force-close.
-	pub(crate) fn update_claims_view<B: Deref, F: Deref, L: Deref>(&mut self, txn_matched: &[&Transaction], requests: Vec<PackageTemplate>, height: u32, broadcaster: &B, fee_estimator: &F, logger: &L)
+	/// `conf_height` represents the height at which the transactions in `txn_matched` were
+	/// confirmed, this does not need to equal the current blockchain tip height, which should be
```
Comment (jkczyz) [inline lightning/src/chain/onchaintx.rs:381]: s/confirmed, this/confirmed. This

File: lightning/src/chain/channelmonitor.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: Right, in that case its being called with reference to a new commitment transaction being broadcast, not one which is confirmed (`new_holder_commitment_transaction`), so I *think* that's correct. Somewhat confusingly, its also called with a 0-height in `provide_payment_preimage` in a similar context, which is "wrong", but AFAICT harmless.

Ultimately it doesn't really matter much for this case, I think, cause its passed through as `height_original` in `PackageTemplate` which is only used to figure out when we can merge packages (which we can never do for the `HolderHTLCOutput`s that are used here), and when we can broadcast (which we just use the CLTV which is fine).

I'll add some comments and fix the 0 value version.

File: lightning/src/chain/channelmonitor.rs (hunk 3 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:2015]: I don't think it matters, but its probably simpler/nicer for users to actually set it there, so I did that.

**Comment by TheBlueMatt** [comment]
Updated to address feedback.

File: lightning/src/chain/channelmonitor.rs (hunk 3 above)
Comment (jkczyz) [reply lightning/src/chain/channelmonitor.rs]: Is this behavior required to fix the bug? Adding it there is equivalent to calling `best_block_updated` after each `transactions_confirmed`. So doesn't this mean calling `best_block_updated` is no longer required when confirming transactions at the same block? Seems it is less simpler for users because it adds a conditional to `Confirm`'s contract, albeit optional.

Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: > Is this behavior required to fix the bug?

No.

> Adding it there is equivalent to calling best_block_updated after each transactions_confirmed. So doesn't this mean calling best_block_updated is no longer required when confirming transactions at the same block?

Yes, I believe so, at least in `ChannelMonitor`, dunno about `ChannelManager`.

> Seems it is less simpler for users because it adds a conditional to Confirm's contract, albeit optional.

I don't see why? It makes a call optional in our current implementation, yes, but we don't have to tell users that.

Comment (jkczyz) [reply lightning/src/chain/channelmonitor.rs]: > I don't see why? It makes a call optional in our current implementation, yes, but we don't have to tell users that.

It seems like surprising behavior. I called X expecting Y but instead Y + Z happened. I guess I don't see why not just drop the comment rather than altering the behavior if it isn't required to fix the bug. At very least it should be a separate commit stating the reason for the change, IMHO.

Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: It definitely feels "more correct" and should happen, IMO. I moved it to a new commit, let me know what you think of the commit docs and if it sufficiently justifies it.

File: lightning/src/chain/channelmonitor.rs [hunk 6]
```diff
@@ -1357,10 +1357,13 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 		// *we* sign a holder commitment transaction, not when e.g. a watchtower broadcasts one of our
 		// holder commitment transactions.
 		if self.broadcasted_holder_revokable_script.is_some() {
-			let (claim_reqs, _) = self.get_broadcasted_holder_claims(&self.current_holder_commitment_tx, 0);
+			// Assume that the broadcasted commitment transaction confirmed in the current best
+			// block. Even if not, its a reasonable metric for the bump criteria on the HTLC
```
Comment (ariard) [inline lightning/src/chain/channelmonitor.rs:1361]: I think this comment is a bit confusing, bumping timer is only function of current height and expiration timelock (see `get_height_timer` in `package.rs`, not of parent transaction conf. So leveraging the best known height as comparison point to refresh timers is the best we can do ?

File: lightning/src/chain/channelmonitor.rs [hunk 7]
```diff
@@ -1747,7 +1750,7 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 						};
 						HolderHTLCOutput::build_accepted(payment_preimage, htlc.amount_msat)
 					};
-				let htlc_package = PackageTemplate::build_package(holder_tx.txid, transaction_output_index, PackageSolvingData::HolderHTLCOutput(htlc_output), height, false, height);
+				let htlc_package = PackageTemplate::build_package(holder_tx.txid, transaction_output_index, PackageSolvingData::HolderHTLCOutput(htlc_output), conf_height, false, conf_height);
```
Comment (ariard) [inline lightning/src/chain/channelmonitor.rs:1753]: I think this is buggy, why we call `build_package` and initialize `soonest_conf_deadline` (i.e the timelock expiration height) to the confirmation or current height ? At least it's already what we do on master (c05347f4), this PR doesn't seem to change anything...

File: lightning/src/chain/channelmonitor.rs [hunk 8]
```diff
@@ -2043,6 +2046,9 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			self.pending_monitor_events.push(MonitorEvent::CommitmentTxBroadcasted(self.funding_info.0));
 			let commitment_tx = self.onchain_tx_handler.get_fully_signed_holder_tx(&self.funding_redeemscript);
 			self.holder_tx_signed = true;
+			// Because we're building a new commitment transaction, we should construct the package
```
Comment (ariard) [inline lightning/src/chain/channelmonitor.rs:2049]: This is confusing too, we never "build" new commitment transaction in the backend, at least we might regenerate one to ensure sanity but never build-and-sign a commitment transaction from scratch ?

**Review by ariard — COMMENTED** [review]
A lot of interactions to think about after a first parse.

File: lightning/src/ln/functional_tests.rs [hunk 9]
```diff
@@ -3001,10 +3001,16 @@ fn do_test_htlc_on_chain_timeout(connect_style: ConnectStyle) {
...
+		// because the original commitment_tx[0] (also spending chan_2.3) has reached
+		// ANTI_REORG_DELAY we really shouldn't be here, and in some connect style scenarios we do
+		// not.
 		let node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().split_off(0);
-		assert_eq!(node_txn.len(), 1);
-		check_spends!(node_txn[0], chan_2.3);
+		if node_txn.len() == 1 {
+			check_spends!(node_txn[0], chan_2.3);
+		} else {
+			assert_eq!(node_txn.len(), 0);
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/functional_tests.rs:3011]: In the old case, a few blocks before we get here we remove the funding tx output from `OnchainTx::claimable_outpoints` (which was originally tracked via `commitment_tx[0]`'s confirmation). Because its not there we can re-add it and re-broadcast our own commitment transaction. In the new case, we skip ahead a few blocks and just call `OnchainTx::update_claims_view` once, which means only at the very end do we remove the funding tx output, after we've already given up broadcasting the commitment tx.

File: lightning/src/chain/channelmonitor.rs [hunk 10]
```diff
@@ -2008,6 +2008,10 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			self.is_paying_spendable_output(&tx, height, &logger);
 		}
 
+		if height > self.best_block.height() {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:2014]: I put it in the doc comment of `block_confirmed` instead.

File: lightning/src/chain/onchaintx.rs [hunk 11]
```diff
@@ -552,7 +555,7 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
 		// Check if any pending claim request must be rescheduled
 		for (first_claim_txid, ref request) in self.pending_claim_requests.iter() {
 			if let Some(h) = request.timer() {
-				if height >= h {
+				if cur_height >= h {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs:558]: Hmmmm, should we swap the ordering of things in `update_claims_view` and first bump RBF transactions and only process the requests after that? I'm a bit hesitant to given it will probably mean a *ton* of test changes.

File: lightning/src/chain/channelmonitor.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:1361]: `get_broadcasted_holder_claims` sets the `PackageTemplate::soonest_conf_deadline` to the height given here, so it *is* the "expiration timelock".

File: lightning/src/chain/channelmonitor.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:1753]: Yea, its "wrong", but only in the sense that there is no "right" answer. We can't RBF-bump the commitment transactions anyway, so its not broken, i don't think, and indeed its the same on master. I'd ideally prefer to clean it up in a separate PR, though I agree its a really poor API.

File: lightning/src/chain/channelmonitor.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: Hmm, semantics, but OK, I tweaked it.

File: lightning/src/chain/channelmonitor.rs [hunk 12]
```diff
@@ -2051,7 +2055,7 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			self.onchain_events_awaiting_threshold_conf.drain(..).collect::<Vec<_>>();
 		let mut onchain_events_reaching_threshold_conf = Vec::new();
 		for entry in onchain_events_awaiting_threshold_conf {
-			if entry.has_reached_confirmation_threshold(height) {
+			if entry.has_reached_confirmation_threshold(self.best_block.height()) {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: There were two issues, this is one, tbh I don't recall exactly which line in `update_claims_view` solved the other one.

**Comment by TheBlueMatt** [comment]
Rebased after merge of the logging dependency. No changes/squashes were done.

**Review by jkczyz — COMMENTED** [review]
Largely looks good though my comprehension of `onchaintx.rs` changes is lacking compared to @ariard.

File: lightning/src/chain/channelmonitor.rs [hunk 13]
```diff
@@ -2004,6 +2004,10 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 			self.is_paying_spendable_output(&tx, height, &logger);
 		}
 
+		if height > self.best_block.height() {
+			self.best_block = BestBlock::new(block_hash, height);
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:2011]: Yea, I think there's a lot of room for improvement here, lets do it in a followup?

[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/chain/channelmonitor.rs:2011]: SGTM

File: lightning/src/chain/channelmonitor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/channelmonitor.rs]: You're right, I don't seep mentions of former bugs over a quick browse of `channel.rs/channelmanager.rs` though for QA purposes what do you think of tracking PR in `do_test_htlc_on_chain_timeout`'s updated comment about connect style scenarios ?

File: lightning/src/chain/channelmonitor.rs (hunk 6 above)
Comment (ariard) [reply lightning/src/chain/channelmonitor.rs:1357]: Yes in this case parent conf height is equal to `PackageTemplate::soonest_conf_deadline` which is equal to "expiration timelock" but my point rather than relying on parent conf height, which have the side-effect of artificially increasing the fee-bump frequency and thus bleed feerate a bit, we should rely on the per-holder-claims timelocks themselves to fulfill `soonest_conf_deadline`. 

Though I don't think this comment needs further action for now and sounds to be related just after.

File: lightning/src/chain/channelmonitor.rs (hunk 7 above)
Comment (ariard) [reply lightning/src/chain/channelmonitor.rs:1749]: Well IMHO the change is as simple as this :

```
diff --git a/lightning/src/chain/channelmonitor.rs b/lightning/src/chain/channelmonitor.rs
index b7f87883..9e6e64e4 100644
--- a/lightning/src/chain/channelmonitor.rs
+++ b/lightning/src/chain/channelmonitor.rs
@@ -1746,7 +1746,7 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
                                                };
                                                HolderHTLCOutput::build_accepted(payment_preimage, htlc.amount_msat)
                                        };
-                               let htlc_package = PackageTemplate::build_package(holder_tx.txid, transaction_output_index, PackageSolvingData::HolderHTLCOutput(htlc_output), conf_height, false, conf_height);
+                               let htlc_package = PackageTemplate::build_package(holder_tx.txid, transaction_output_index, PackageSolvingData::HolderHTLCOutput(htlc_output), htlc.cltv_expiry, false, conf_height);
                                claim_requests.push(htlc_package);
                        }
                }
```

Though doesn't break any test on top of this branch, so likely need a test to demonstrate the fee-bumping policy change. Tracked as #982 and self-assigned. Note, it might duplicate `htlc.cltv_expiry` usage both at the package-level and at the output-level but i don't think that's a redundance as in the first case we would use for fee-bumping scheduling and in the second to generate the correct `witnessScript` ?

File: lightning/src/ln/functional_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/functional_tests.rs:3011]: > Because its not there we can re-add it and re-broadcast our own commitment transaction.

This is the weirdness I didn't understand and AFAICT it sounds an issue with our already-existent `update_claim_view` logic where we might have a "height" collision between maturity of `OnchainEvent::Claim` for `commitment_tx[0]` and hitting a `should_broadcast()==true` generating a new request for the same commitment.

Adding the following diff

```
diff --git a/lightning/src/chain/onchaintx.rs b/lightning/src/chain/onchaintx.rs
index 7e616ef2..4e1c1c5d 100644
--- a/lightning/src/chain/onchaintx.rs
+++ b/lightning/src/chain/onchaintx.rs
@@ -536,6 +536,7 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
                                                // been aggregated in a single tx and claimed so atomically
                                                if let Some(request) = self.pending_claim_requests.remove(&claim_request) {
                                                        for outpoint in request.outpoints() {
+                                                               log_trace!(logger, "Removing claiming outpoint for {}", outpoint.txid);
                                                                self.claimable_outpoints.remove(&outpoint);
                                                        }
                                                }
```                                                

And obtaining the following log, confirm that we might remove a tracked output to re-add just after as a new request ? (see `24647a3792b08dbcd7c9516ae5118fce200b2a11c49cbc496b48da466b125e3e`)

```

TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 1925] New best block 88fb36eb50c852280f8b85410462252dff993bbf50a3e752dec3f8a54b978c41 at height 206                                                                                              
INFO  node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2250] Force-closing channel due to outbound HTLC timeout, HTLC expiry is 81                                                                                                                      
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2093] Descriptor DynamicOutputP2WPKH 012b0c6223264225513eb7fd1661b3e04e6076f8c0868b48565ba85d9c3c24c4:2 marked for spending has got enough confirmations to be passed upstream                   
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2085] HTLC 75877bb41d393b5fb8455ce60ecd8dda001d06316496b14dfa7f895656eeca4a failure update has got enough confirmations to be passed upstream                                                    
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2093] Descriptor StaticOutput 4ee8b85ba72a1a1beb201438e8ec8903ce0acf2e48161e5e61c9e331d048fa70:0 marked for spending has got enough confirmations to be passed upstream                          
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 385] Updating claims view at height 206 with 0 matched transactions and 2 claim requests                                                                                                                   
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 394] Ignoring second claim for outpoint 24647a3792b08dbcd7c9516ae5118fce200b2a11c49cbc496b48da466b125e3e:0, already registered its claiming request                                                        
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 394] Ignoring second claim for outpoint b00e0a6ecc3ad8958be2552cf1eefc75232be2c68ec4668e14f6c784d5d050e3:0, already registered its claiming request                                                        
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 539] Removing claiming outpoint for 24647a3792b08dbcd7c9516ae5118fce200b2a11c49cbc496b48da466b125e3e                                                                                                       
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 539] Removing claiming outpoint for 012b0c6223264225513eb7fd1661b3e04e6076f8c0868b48565ba85d9c3c24c4                                                                                                       
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 563] Bumping 0 candidates
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 1925] New best block 88fb36eb50c852280f8b85410462252dff993bbf50a3e752dec3f8a54b978c41 at height 206                                                                                              
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 385] Updating claims view at height 206 with 0 matched transactions and 0 claim requests                                                                                                                   
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 563] Bumping 0 candidates
TRACE node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 1963] Block 88fb36eb50c852280f8b85410462252dff993bbf50a3e752dec3f8a54b978c41 at height 206 connected with 0 txn matched                                                                          
INFO  node 1 [lightning::chain::channelmonitor : lightning/src/chain/channelmonitor.rs, 2250] Force-closing channel due to outbound HTLC timeout, HTLC expiry is 81                                                                                                                      
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 385] Updating claims view at height 206 with 0 matched transactions and 2 claim requests                                                                                                                   
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 413] Test if outpoint can be aggregated with expiration 206 against 218                                                                                                                                    
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 394] Ignoring second claim for outpoint b00e0a6ecc3ad8958be2552cf1eefc75232be2c68ec4668e14f6c784d5d050e3:0, already registered its claiming request                                                        
TRACE node 1 [lightning::chain::package : lightning/src/chain/package.rs, 615] Adding claiming input for outpoint 24647a3792b08dbcd7c9516ae5118fce200b2a11c49cbc496b48da466b125e3e:0                                                                                                     
TRACE node 1 [lightning::chain::package : lightning/src/chain/package.rs, 616] Finalized transaction b00e0a6ecc3ad8958be2552cf1eefc75232be2c68ec4668e14f6c784d5d050e3 ready to broadcast                                                                                                 
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 444] Registering claiming request for 24647a3792b08dbcd7c9516ae5118fce200b2a11c49cbc496b48da466b125e3e:0                                                                                                   
TRACE node 1 [lightning::chain::onchaintx : lightning/src/chain/onchaintx.rs, 448] Broadcasting onchain commitment tx with txid b00e0a6ecc3ad8958be2552cf1eefc75232be2c68ec4668e14f6c784d5d050e3 

```

File: lightning/src/chain/onchaintx.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/onchaintx.rs:561]: See above comment about height collision, I would favor this ordering swap to avoid re-broadcasting. I believe, post-anchor that kind of defaulting internal API issue might force us to reserve fee-bumping reserve for nothing ?

Though it's not related to this PR, and we can keep thinking about it.

File: lightning/src/chain/channelmonitor.rs [hunk 14]
```diff
@@ -2200,7 +2219,7 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 		false
 	}
 
-	fn would_broadcast_at_height<L: Deref>(&self, height: u32, logger: &L) -> bool where L::Target: Logger {
+	fn should_broadcast<L: Deref>(&self, logger: &L) -> bool where L::Target: Logger {
```
Comment (ariard) [inline lightning/src/chain/channelmonitor.rs:2222]: nit: as you retouched the name, maybe it could be called `should_broadcast_holder_commitment`? imho marginally improving readability of this confusing part of the codebase where a wide array of transactions is constantly processed.

**Review by ariard — COMMENTED** [review]
Notwithstanding already-existent issues mentioned in comment and we should address latter, this PR sounds good to me.

File: lightning/src/ln/functional_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/functional_tests.rs:3011]: >  This is the weirdness I didn't understand and AFAICT it sounds an issue with our already-existent update_claim_view logic where we might have a "height" collision between maturity of OnchainEvent::Claim for commitment_tx[0] and hitting a should_broadcast()==true generating a new request for the same commitment.

Its not height, its the previous outpoint itself. The `claimable_outpoints` map is indexed by previous outpoint, and in this case we've already registered that we're tracking a spend of the same outpoint, so we don't add a new one. This seems correct to me, if race-y.

File: lightning/src/chain/channelmonitor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs]: Hmm, I generally just `git blame` and then look at the commit(s)/check github to see what PR it came from. We could add PR #s too, but I'm a bit hesitant to have the code explicitly reference the PR it was in. But if you feel strongly I can add it.

File: lightning/src/chain/onchaintx.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs:561]: I'm a bit confused about that comment, but I'd think the solution is to track outputs which were on chain and spent and *never* re-spend them, instead of re-spending them after ANTI_REORG_DELAY.

File: lightning/src/chain/channelmonitor.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:1749]: Ah! Duh, indeed, I was somehow confusing the htlc package for the commitment tx package. I updated it but agree we need more tests here.

File: lightning/src/chain/channelmonitor.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/chain/channelmonitor.rs:1357]: I'm still a bit confused here, but I'll leave it for a followup.

File: lightning/src/ln/functional_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/functional_tests.rs:3011]: Hmmmm i still think it's a weird height-race because note in the log `channelmonitor` is processing twice height 206 ? First time we remove the commitment's outpoint from `OnchainTxHandler.claimable_outpoints` then at second processing we add a new request for the _already-confirmed_ commitment transaction in `OnchainTxHandler.claimable_outpoints` as the redundant-claim replay protection L393 doesn't play out ?

File: lightning/src/chain/channelmonitor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/channelmonitor.rs]: Well i'm a huge fan of good post-mortem tracking, kind of in the long-term of #573 :) Making a mental note to add to next agenda meeting as we have all differing opinions there.

File: lightning/src/chain/onchaintx.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/chain/onchaintx.rs:561]: Yeah that would works too to blacklist confirmed utxos so i think it's more an issue on how duplicate block processing is interfering with `update_claim_view()` list of calls.

**Comment by ariard** [comment]
Code Review ACK 5e51f84

Thanks for taking the latest suggestion!

File: lightning/src/ln/functional_tests.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/functional_tests.rs:3011]: Right, I totally agree *that* is a (very minor) bug, but this PR just removes that bug depending on the connection style, so I don't think it needs to get fixed here :).

File: lightning/src/chain/onchaintx.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/chain/onchaintx.rs:561]: Oh, hmm, I guess I just misread your above comment. Let me try the ordering swap in a new PR? I don't really want to hold this/0.0.99 up to redo a ton of tests.

**Review by jkczyz — COMMENTED** [review]
ACK 5e51f84b571af51d28e543fd14f33034dcbb9771

**Review by jkczyz — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed with no diff, will take after CI:
```
$ git diff-tree -U1 5e51f84b 412cc9f0
$
```
